### ISSUE TITLE
"null value in column id value" Error when restoring database on backup VM

### ISSUE THREAD

**Issue by mdhar1980** [tripal/tripal#989]
Hello!

We have a Tripal (v 2.1) installed and running on a production server, and had copied over files and used archive/restore of the database to copy everything onto a VM to serve as backup/development server. The sites look identical when visited as a user. However, when signing into the VM/development version, an error appears and freezes the site if you click on any admin menu item other than "Add Content." It reads as follows:

> PDOException: SQLSTATE[23502]: Not null violation: 7 ERROR: null value in column "item_id" violates not-null constraint DETAIL: Failing row contains (null, update_fetch_tasks, \x613a383a7b733a343a226e616d65223b733a363a2264727570616c223b733a..., 0, 1566568544).: INSERT INTO queue (name, data, created) VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2); Array ( ) in SystemQueue->createItem() (line 221 of /var/www/html/modules/system/system.queue.inc).

I haven't been able to resolve this, so I was wondering if anyone had encountered it and found a solution. I realize this may be PostGRES-related, so I'm also planning on posting on such a board, but thought I'd start here, since it is on a Tripal/Drupal site.

I used drush commands for backing up the database and "restoring" onto the VM server:
drush sql-dump > /share/Backups/database-backup_7-18-19.sql
`drush sql-connect` < ~/Backups/server-database-backup_7-18-19.sql 

Site version info:
Tripal Version: 7.x-2.1
Drupal Version: 767
PostgreSQL Version: PostgreSQL 10.10 (Ubuntu 10.10-0ubuntu0.18.04.1)
PHP Version: 7.2.19-0ubuntu0.18.04.2

Thanks for any help!
--Michael

**Comment by spficklin** [comment]
Hi @mdhar1980 . If you look at the table structure for the Drupal `queue` table, you'll see that the `item_id` column has a default value that gets populated by the `queue_item_id_seq` sequence.  

![image](https://user-images.githubusercontent.com/1719352/63866323-e1494280-c967-11e9-8ece-8d1fdb49ccb0.png)

 If Drupal is trying to insert a value into that table (as indicated in the error message you show above) and is not supplying the `item_id`, then it is relying on the sequence to auto-populate it (i.e. a primary key).  

Therefore, I would suspect that when you dumped and restored the database you might have missed the sequences used to auto-populate the primary keys of tables?

**Comment by mdhar1980** [comment]
Hi, Stephen,

Thanks for your response! 

For the dump and restore, I simply used those drush commands. Unless I misused them, would that then mean those methods miss some items? Is there a way to ensure dump/restore doesn't miss primary keys?

Thanks again,
Michael

**Comment by spficklin** [comment]
What drush commands did you use?

In the past, I've used this command to dump the database:

`pg_dump --no-owner  -h [host] -U [user] -W -f [database].sql [database] `

And this to restore:

`psql -U [user] -h [host] -d [database] -f [database].sql`

**Comment by mdhar1980** [comment]
Hi, Stephen,

Thanks, I'll give those a try.

I used drush commands:

drush sql-dump > /share/Backups/database-backup_7-18-19.sql
`drush sql-connect` < ~/Backups/server-database-backup_7-18-19.sql

**Comment by spficklin** [comment]
Yeah, sorry. I see you posted those in your original message above. Hopefully the PostgreSQL dump command work for you.

**Comment by mdhar1980** [comment]
There's some automatic formatting going on here, but there are back tics around "drush sql-connect"

**Comment by mdhar1980** [comment]
Hi, reporting back: Unfortunately, those backup/restore commands resulted in the same error:

> PDOException: SQLSTATE[23502]: Not null violation: 7 ERROR: null value in column "item_id" violates not-null constraint DETAIL: Failing row contains (null, update_fetch_tasks, \x613a383a7b733a343a226e616d65223b733a31363a2274726970616c5f646f..., 0, 1567352152).: INSERT INTO queue (name, data, created) VALUES (:db_insert_placeholder_0, :db_insert_placeholder_1, :db_insert_placeholder_2); Array ( ) in SystemQueue->createItem() (line 221 of /var/www/html/modules/system/system.queue.inc).

**Comment by mdhar1980** [comment]
Hi,

I was able to resolve this by upgrading the postgres version on our backup VM. (It had been set up earlier, and so I realized it had postgresql v 9.5, whereas the production server has v 10.8). 

I upgraded to postgres 10 (using the method described here: https://gist.github.com/mdesantis/ebab306c0bf13f785b93e7894ca0cde2)

The existing database, updated from the 9.5 version, still had the same error -- probably because the db was created with an error when I dumped into the 9.5 version from the 10.8 version. However, I created a new database, dumped into that, and this resolved the issue.

Thanks!
